google-cloud-resourcemanager/src/main/java/com/google/cloud/resourcemanager/spi/v1beta1/HttpResourceManagerRpc.java

```diff
@@ -16,58 +16,198 @@
+import org.threeten.bp.Duration;
 
 public class HttpResourceManagerRpc implements ResourceManagerRpc {
 
-  private final Cloudresourcemanager resourceManager;
+  private static final JsonFactory JSON_FACTORY =
+      new com.google.api.client.json.jackson.JacksonFactory();
+
+  // See doc of create() for more details:
+  // https://developers.google.com/resources/api-libraries/documentation/cloudresourcemanager/v1/java/latest/com/google/api/services/cloudresourcemanager/CloudResourceManager.Projects.html#create(com.google.api.services.cloudresourcemanager.model.Project)
+  private static final RetrySettings CREATE_RETRY_SETTINGS =
+      RetrySettings.newBuilder()
+          // SLO permits 30s at 90th percentile, double it for total limit
+          .setTotalTimeout(Duration.ofMinutes(1))
+          .setInitialRetryDelay(Duration.ofSeconds(5))
+          .setRetryDelayMultiplier(1.5)
+          // Observed P95 latency is 11s. We probably shouldn't sleep longer than this.
+          .setMaxRetryDelay(Duration.ofSeconds(11))
+          .setJittered(false)
+          .setInitialRpcTimeout(Duration.ofSeconds(5))
+          .setMaxRpcTimeout(Duration.ofSeconds(5))
+          .build();
+
+  private static final ResultRetryAlgorithm<Operation> OPERATION_HANDLER =
+      new ResultRetryAlgorithm<Operation>() {
+        @Override
+        public TimedAttemptSettings createNextAttempt(
+            Throwable prevThrowable, Operation prevResponse, TimedAttemptSettings prevSettings) {
+          return null;
+        }
+
+        @Override
+        public boolean shouldRetry(Throwable prevThrowable, Operation prevOp)
+            throws CancellationException {
+          return prevThrowable != null || prevOp.getDone() == null || prevOp.getDone() == false;
+        }
+      };
+
+  private final CloudResourceManager resourceManager;
+  private final ApiClock clock;
 
   public HttpResourceManagerRpc(ResourceManagerOptions options) {
     HttpTransportOptions transportOptions = (HttpTransportOptions) options.getTransportOptions();
     HttpTransport transport = transportOptions.getHttpTransportFactory().create();
     HttpRequestInitializer initializer = transportOptions.getHttpRequestInitializer(options);
     resourceManager =
-        new Cloudresourcemanager.Builder(transport, new JacksonFactory(), initializer)
+        new CloudResourceManager.Builder(transport, new JacksonFactory(), initializer)
             .setRootUrl(options.getHost())
             .setApplicationName(options.getApplicationName())
             .build();
+    clock = options.getClock();
   }
 
   private static ResourceManagerException translate(IOException exception) {
     return new ResourceManagerException(exception);
   }
 
+  private static ResourceManagerException translate(Status status) {
+    int code;
+    // reference: https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto
+    switch (status.getCode()) {
```

Comment:
Can this be just a map of `statusCode => code`? Having such a long and narrow switch statement hurts my eyes. Having a map at least makes it shorter:

The constant in the class:
```java
    private static final ImmutableMap<Integer, Integer> ERROR_CODES_MAP =
        ImmutableMap.builder()
            .put(0, 200)
            .put(1, 499)
            .put(2, 500)
            .put(3, 400)
            .put(4, 504)
            .put(5, 404)
            .put(6, 409)
            .put(7, 403)
            .put(16, 401)
            .put(8, 429)
            .put(9, 400)
            .put(10, 409)
            .put(11, 400)
            .put(12, 501)
            .put(13, 500)
            .put(14, 503)
            .put(15, 500)
            .build();
```

Then in this method:
```java
    code = ERROR_CODES_MAP.getOrDefault(status.getCode(), BaseHttpServiceException.UNKNOWN_CODE);
```

Reply:
Good idea, done.